sbat.c

```diff
@@ -117,4 +118,191 @@ parse_sbat(char *sbat_base, size_t sbat_size, struct sbat *sbat)
+			FreePool((CHAR8 *)entry->component_generation);
+		FreePool(entry);
+	}
+}
+
+EFI_STATUS
+verify_sbat(struct sbat *sbat, list_t *sbat_entries)
+{
+	unsigned int i;
+	list_t *pos = NULL;
+	EFI_STATUS efi_status = EFI_SUCCESS;
+	struct sbat_entry *entry;
+	struct sbat_var *sbat_var_entry;
+
+	if (!sbat_entries || sbat_empty(sbat_entries)) {
+		dprint(L"SBAT variable not present or malformed\n");
+		return EFI_INVALID_PARAMETER;
+	}
+
+	for (i = 0; i < sbat->size; i++) {
+		entry = sbat->entries[i];
+		for_each_sbat_var(pos, sbat_entries)
+		{
+			sbat_var_entry = list_entry(pos, struct sbat_var, list);
+			efi_status = verify_single_entry(entry, sbat_var_entry);
+			if (EFI_ERROR(efi_status))
+				return efi_status;
+		}
+	}
+
+	dprint(L"all entries from SBAT section verified\n");
+	clean_up_vars(sbat_entries);
+	return efi_status;
+}
+
+static BOOLEAN
+is_utf8_bom(CHAR8 *buf, size_t bufsize)
+{
+	unsigned char bom[] = { 0xEF, 0xBB, 0xBF };
+
+	return CompareMem(buf, bom, MIN(sizeof(bom), bufsize)) == 0;
+}
+
+static struct sbat_var *
+new_entry(const CHAR8 *comp_name, const CHAR8 *comp_gen)
+{
+	struct sbat_var *new_entry = AllocatePool(sizeof(*new_entry));
+
+	if (!new_entry)
+		return NULL;
+	INIT_LIST_HEAD(&new_entry->list);
+	new_entry->component_name = comp_name;
+	new_entry->component_generation = comp_gen;
+
+	return new_entry;
+}
+
+EFI_STATUS
+add_entry(list_t *list, const CHAR8 *comp_name, const CHAR8 *comp_gen)
+{
+	struct sbat_var *new;
+
+	new = new_entry(comp_name, comp_gen);
+	if (!new)
+		return EFI_OUT_OF_RESOURCES;
+
+	list_add_tail(&new->list, list);
+	return EFI_SUCCESS;
+}
+
+EFI_STATUS
+parse_sbat_var(list_t *entries)
+{
+	UINT8 *data = 0;
+	UINTN datasize, i;
+	EFI_STATUS efi_status;
+	INIT_LIST_HEAD(entries);
+	char delim;
+
+	if (!entries)
+		return EFI_INVALID_PARAMETER;
+
+	efi_status = get_variable(L"SBAT", &data, &datasize, SHIM_LOCK_GUID);
+	if (EFI_ERROR(efi_status)) {
+		LogError(L"Failed to read SBAT variable\n",
+			 efi_status);
+		return efi_status;
+	}
+
+	CHAR8 *start = (CHAR8 *)data;
+	CHAR8 *end = (CHAR8 *)data + datasize;
+	if (is_utf8_bom(start, datasize))
+		start += 3;
+
+	dprint(L"SBAT variable data:\n");
+
+	while (start[0] != '\0') {
+		const CHAR8 *fields[2] = {
+			NULL,
+		};
+		for (i = 0; i < 3; i++) {
+			const CHAR8 *tmp;
+			/*
+			 * on third iteration we check if we had extra stuff on line while parsing
+			 * component_name. If delimeter on 2nd iteration was ',', this means that
+			 * we have comments after component_name. get_sbat_field in this if condition
+			 * parses comments, if they are present and drops them.
+			 */
+			if (i == 2 && start) {
+				if (delim == ',') {
+					start = get_sbat_field(start, end, &tmp,
+					                       '\n');
+				}
+				break;
+			}
+			delim = ',';
+			/* we do not want to jump to next line and grab stuff from that
+			 */
+			if ((strchrnula(start, '\n') - start + 1) <=
+			    (strchrnula(start, ',') - start + 1)) {
```

Comment:
If this condition is true on the first iteration of the inner loop, should the payload be considered malformed?

Reply:
should be fixed